File: java/arrays/medium/ProductOfArrayExceptSelf.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int length = nums.length;
        
        int[]left = new int [length];
        int[]right = new int [length];
        int[]answer = new int [length];
        
        left[0] = 1;
        
        for (int i = 1; i < length; i++) {
            left[i] = left[i-1] * nums[i-1];
        }
        
        right[length-1] = 1;
        for (int i = length -2; i>=0; i--) {
            right[i] = right[i+1] * nums[i+1];
        }
        
        for (int i = 0; i< length; i++) {
            answer[i] = right[i] * left[i];
        }
        return answer;
    }
}

File: Python/Arrays/Medium/productOfArrayExceptSelf.py
class Solution(object):
    def productOfArrayExceptSelf(self, nums):
        # find length of nums and init result array with length 
        length = len(nums)
        result = [0] * length 

        # init 0th index of result to 1
        result[0] = 1 

        # loop through nums from left -> right and store product in result
        for i in range(1, length):
            result[i] = nums[i - 1] * result[i - 1]
        
        # init right to 1 
        right = 1

        # loop through nums from right to left and calculate product 
        for i in reversed(range(length)):
            result[i] = result[i] * right
            right *= nums[i]
        return result

#Time comp: O(n)
#Space comp: O(1)

